--- a/plugins/log/src/lib.rs
+++ b/plugins/log/src/lib.rs
@@ -750,9 +750,9 @@
                         );
                     }
                 }),
-                #[cfg(desktop)]
+                #[cfg(any(desktop, target_env = "ohos"))]
                 TargetKind::Stdout => std::io::stdout().into(),
-                #[cfg(desktop)]
+                #[cfg(any(desktop, target_env = "ohos"))]
                 TargetKind::Stderr => std::io::stderr().into(),
                 TargetKind::Folder { path, file_name } => {
                     if !path.exists() {
