diff --git a/Cargo.toml b/Cargo.toml
index 6f2a024..4e2a1d9 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -109,6 +109,10 @@ ndk-context = "0.1"
 percent-encoding = "2.3"
 tao-macros = { version = "0.1.4", path = "./tao-macros" }
 
+[target."cfg(target_env = \"ohos\")".dependencies]
+openharmony-ability = { git = "https://github.com/harmony-contrib/openharmony-ability.git", rev = "295a276a699ba2352addc69b8cacb6acd3be6ab1" }
+openharmony-ability-derive = { git = "https://github.com/harmony-contrib/openharmony-ability.git", rev = "295a276a699ba2352addc69b8cacb6acd3be6ab1" }
+
 [target."cfg(any(target_os = \"ios\", target_os = \"macos\"))".dependencies]
 objc2 = "0.6"
 block2 = "0.6"
@@ -165,7 +169,7 @@ objc2-foundation = { version = "0.3", default-features = false, features = [
   "NSProcessInfo",
 ] }
 
-[target."cfg(any(target_os = \"linux\", target_os = \"dragonfly\", target_os = \"freebsd\", target_os = \"openbsd\", target_os = \"netbsd\"))".dependencies]
+[target."cfg(all(any(target_os = \"linux\", target_os = \"dragonfly\", target_os = \"freebsd\", target_os = \"openbsd\", target_os = \"netbsd\"), not(target_env = \"ohos\")))".dependencies]
 gtk = "0.18"
 gdkx11-sys = { version = "0.18", optional = true }
 gdkwayland-sys = "0.18.0"
diff --git a/src/event.rs b/src/event.rs
index 14325dc..17397e0 100644
--- a/src/event.rs
+++ b/src/event.rs
@@ -797,7 +797,7 @@ pub struct KeyEvent {
   pub(crate) platform_specific: platform_impl::KeyEventExtra,
 }
 
-#[cfg(not(any(target_os = "android", target_os = "ios")))]
+#[cfg(not(any(target_os = "android", target_os = "ios", target_env = "ohos")))]
 impl KeyEvent {
   /// Identical to `KeyEvent::text` but this is affected by <kbd>Ctrl</kbd>.
   ///
@@ -821,7 +821,7 @@ impl KeyEvent {
   }
 }
 
-#[cfg(any(target_os = "android", target_os = "ios"))]
+#[cfg(any(target_os = "android", target_os = "ios", target_env = "ohos"))]
 impl KeyEvent {
   /// Identical to `KeyEvent::text`.
   pub fn text_with_all_modifiers(&self) -> Option<&str> {
diff --git a/src/event_loop.rs b/src/event_loop.rs
index eb73dbb..4472d1d 100644
--- a/src/event_loop.rs
+++ b/src/event_loop.rs
@@ -216,6 +216,7 @@ impl<T> EventLoop<T> {
   ///   disconnects.
   ///
   /// [`ControlFlow`]: crate::event_loop::ControlFlow
+  #[cfg(not(target_env = "ohos"))]
   #[inline]
   pub fn run<F>(self, event_handler: F) -> !
   where
@@ -224,6 +225,16 @@ impl<T> EventLoop<T> {
     self.event_loop.run(event_handler)
   }
 
+  /// For OpenHarmony, the event loop is not blocking, so we need to return a unit type.
+  #[cfg(target_env = "ohos")]
+  #[inline]
+  pub fn run<F>(self, event_handler: F) -> ()
+  where
+    F: 'static + FnMut(Event<'_, T>, &EventLoopWindowTarget<T>, &mut ControlFlow),
+  {
+    self.event_loop.run(event_handler)
+  }
+
   /// Creates an `EventLoopProxy` that can be used to dispatch user events to the main event loop.
   pub fn create_proxy(&self) -> EventLoopProxy<T> {
     EventLoopProxy {
@@ -306,7 +317,10 @@ impl<T> EventLoopWindowTarget<T> {
   /// - **iOS / Android:** Unsupported.
   #[inline]
   pub fn set_progress_bar(&self, _progress: ProgressBarState) {
-    #[cfg(any(target_os = "linux", target_os = "macos"))]
+    #[cfg(all(
+      any(target_os = "linux", target_os = "macos"),
+      not(target_env = "ohos")
+    ))]
     self.p.set_progress_bar(_progress)
   }
 
@@ -317,14 +331,17 @@ impl<T> EventLoopWindowTarget<T> {
   /// - **iOS / Android:** Unsupported.
   #[inline]
   pub fn set_theme(&self, _theme: Option<Theme>) {
-    #[cfg(any(
-      windows,
-      target_os = "linux",
-      target_os = "dragonfly",
-      target_os = "freebsd",
-      target_os = "netbsd",
-      target_os = "openbsd",
-      target_os = "macos",
+    #[cfg(all(
+      any(
+        windows,
+        target_os = "linux",
+        target_os = "dragonfly",
+        target_os = "freebsd",
+        target_os = "netbsd",
+        target_os = "openbsd",
+        target_os = "macos",
+      ),
+      not(target_env = "ohos")
     ))]
     self.p.set_theme(_theme)
   }
diff --git a/src/keyboard.rs b/src/keyboard.rs
index 8628f0f..84ccf4c 100644
--- a/src/keyboard.rs
+++ b/src/keyboard.rs
@@ -193,6 +193,7 @@ pub enum NativeKeyCode {
   /// This is the android "key code" of the event as returned by
   /// `KeyEvent.getKeyCode()`
   Android(i32),
+  Ohos(i32),
 }
 
 /// Represents the code of a physical key.
diff --git a/src/platform/linux.rs b/src/platform/linux.rs
index 2f0603c..6421d5a 100644
--- a/src/platform/linux.rs
+++ b/src/platform/linux.rs
@@ -2,4 +2,4 @@
 // Copyright 2021-2023 Tauri Programme within The Commons Conservancy
 // SPDX-License-Identifier: Apache-2.0
 
-#![cfg(target_os = "linux")]
+#![cfg(all(target_os = "linux", not(target_env = "ohos")))]
diff --git a/src/platform/mod.rs b/src/platform/mod.rs
index 2ca20b3..31669f3 100644
--- a/src/platform/mod.rs
+++ b/src/platform/mod.rs
@@ -23,6 +23,7 @@ pub mod android;
 pub mod ios;
 pub mod linux;
 pub mod macos;
+pub mod ohos;
 pub mod run_return;
 pub mod unix;
 pub mod windows;
diff --git a/src/platform/ohos.rs b/src/platform/ohos.rs
new file mode 100644
index 0000000..c18665c
--- /dev/null
+++ b/src/platform/ohos.rs
@@ -0,0 +1,94 @@
+//! # OpenHarmony
+//!
+//! The OpenHarmony backend builds on (and exposes types from) the [`ohos-rs`](https://docs.rs/ohos-rs/) crate.
+//!
+//! Native OpenHarmony applications need some form of "glue" crate that is responsible
+//! for defining the main entry point for your Rust application as well as tracking
+//! various life-cycle events and synchronizing with the main thread.
+//!
+//! Winit uses the [openharmony-ability](https://docs.rs/openharmony-ability/) as a
+//! glue crate.
+//!
+
+#![cfg(target_env = "ohos")]
+
+use crate::event_loop::{EventLoop, EventLoopBuilder};
+use crate::window::{Window, WindowAttributes};
+use openharmony_ability::{Configuration, OpenHarmonyApp, Rect};
+
+/// Additional methods on [`EventLoop`] that are specific to OpenHarmony.
+pub trait EventLoopExtOpenHarmony {}
+
+impl<T> EventLoopExtOpenHarmony for EventLoop<T> {}
+
+/// Additional methods on [`Window`] that are specific to OpenHarmony.
+pub trait WindowExtOpenHarmony {
+  fn content_rect(&self) -> Rect;
+
+  fn config(&self) -> Configuration;
+}
+
+impl WindowExtOpenHarmony for Window {
+  fn content_rect(&self) -> Rect {
+    self.window.content_rect()
+  }
+
+  fn config(&self) -> Configuration {
+    self.window.config()
+  }
+}
+
+/// Additional methods on [`WindowAttributes`] that are specific to OpenHarmony.
+pub trait WindowAttributesExtOpenHarmony {}
+
+impl WindowAttributesExtOpenHarmony for WindowAttributes {}
+
+pub trait EventLoopBuilderExtOpenHarmony {
+  /// Associates the [`OpenHarmonyApp`] that was passed to `openharmony-ability::ability` with the event loop
+  ///
+  /// This must be called on OpenHarmony since the [`OpenHarmonyApp`] is not global state.
+  fn with_openharmony_app(&mut self, app: OpenHarmonyApp) -> &mut Self;
+}
+
+impl<T> EventLoopBuilderExtOpenHarmony for EventLoopBuilder<T> {
+  fn with_openharmony_app(&mut self, app: OpenHarmonyApp) -> &mut Self {
+    self.platform_specific.openharmony_app = Some(app);
+    self
+  }
+}
+
+/// Re-export of the `openharmony-ability` API
+///
+/// Winit re-exports the `openharmony-ability` API for convenience so that most
+/// applications can rely on the Winit crate to resolve the required version of
+/// `openharmony-ability` and avoid any chance of a conflict between Winit and the
+/// application crate.
+///
+///
+/// For compatibility applications should then import the [`OpenHarmonyApp`] type for
+/// their `init(app: OpenHarmonyApp)` function and use `openharmony-ability-derive` to
+/// implement entry like:
+/// ```rust
+/// #[cfg(target_env = "ohos")]
+/// use winit::platform::ohos::ability::OpenHarmonyApp;
+/// use openharmony_ability_derive::ability;
+///
+/// #[ability]
+/// fn init(app: OpenHarmonyApp) {
+///     // ...
+/// }
+/// ```
+pub mod ability {
+  #[doc(no_inline)]
+  pub use openharmony_ability::*;
+
+  #[doc(no_inline)]
+  pub use openharmony_ability_derive::*;
+
+  #[doc(hidden)]
+  pub struct Rect;
+  #[doc(hidden)]
+  pub struct ConfigurationRef;
+  #[doc(hidden)]
+  pub struct OpenHarmonyApp;
+}
diff --git a/src/platform/unix.rs b/src/platform/unix.rs
index c25b89f..4534ea6 100644
--- a/src/platform/unix.rs
+++ b/src/platform/unix.rs
@@ -2,12 +2,15 @@
 // Copyright 2021-2023 Tauri Programme within The Commons Conservancy
 // SPDX-License-Identifier: Apache-2.0
 
-#![cfg(any(
-  target_os = "linux",
-  target_os = "dragonfly",
-  target_os = "freebsd",
-  target_os = "netbsd",
-  target_os = "openbsd"
+#![cfg(all(
+  any(
+    target_os = "linux",
+    target_os = "dragonfly",
+    target_os = "freebsd",
+    target_os = "netbsd",
+    target_os = "openbsd"
+  ),
+  not(target_env = "ohos")
 ))]
 
 #[cfg(feature = "x11")]
diff --git a/src/platform_impl/mod.rs b/src/platform_impl/mod.rs
index e8b62af..9e88d2e 100644
--- a/src/platform_impl/mod.rs
+++ b/src/platform_impl/mod.rs
@@ -5,12 +5,15 @@
 #[cfg(target_os = "windows")]
 #[path = "windows/mod.rs"]
 mod platform;
-#[cfg(any(
-  target_os = "linux",
-  target_os = "dragonfly",
-  target_os = "freebsd",
-  target_os = "netbsd",
-  target_os = "openbsd"
+#[cfg(all(
+  any(
+    target_os = "linux",
+    target_os = "dragonfly",
+    target_os = "freebsd",
+    target_os = "netbsd",
+    target_os = "openbsd"
+  ),
+  not(target_env = "ohos")
 ))]
 #[path = "linux/mod.rs"]
 mod platform;
@@ -23,6 +26,9 @@ mod platform;
 #[cfg(target_os = "ios")]
 #[path = "ios/mod.rs"]
 mod platform;
+#[cfg(target_env = "ohos")]
+#[path = "ohos/mod.rs"]
+mod platform;
 
 pub use platform::*;
 
@@ -36,5 +42,6 @@ pub use platform::*;
   not(target_os = "freebsd"),
   not(target_os = "netbsd"),
   not(target_os = "openbsd"),
+  not(target_env = "ohos")
 ))]
 compile_error!("The platform you're compiling for is not supported by tao");
diff --git a/src/platform_impl/ohos/keycodes.rs b/src/platform_impl/ohos/keycodes.rs
new file mode 100644
index 0000000..90d4e45
--- /dev/null
+++ b/src/platform_impl/ohos/keycodes.rs
@@ -0,0 +1,377 @@
+use openharmony_ability::xcomponent::KeyCode as Keycode;
+
+use crate::keyboard::{Key, KeyLocation, NativeKeyCode};
+
+pub fn to_logical(keycode: Keycode) -> Key<'static> {
+  use openharmony_ability::xcomponent::KeyCode::*;
+
+  let native = NativeKeyCode::Ohos(i32::from(keycode));
+
+  match keycode {
+    // Using `BrowserHome` instead of `GoHome` according to
+    // https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key/Key_Values
+    Home => Key::BrowserHome,
+    Back => Key::BrowserBack,
+
+    //-------------------------------------------------------------------------------
+    // These should be redundant because they should have already been matched
+    // as `KeyMapChar::Unicode`, but also matched here as a fallback
+    Key0 => Key::Unidentified(native),
+    Key1 => Key::Unidentified(native),
+    Key2 => Key::Unidentified(native),
+    Key3 => Key::Unidentified(native),
+    Key4 => Key::Unidentified(native),
+    Key5 => Key::Unidentified(native),
+    Key6 => Key::Unidentified(native),
+    Key7 => Key::Unidentified(native),
+    Key8 => Key::Unidentified(native),
+    Key9 => Key::Unidentified(native),
+    Star => Key::Unidentified(native),
+    Pound => Key::Unidentified(native),
+    A => Key::Unidentified(native),
+    B => Key::Unidentified(native),
+    C => Key::Unidentified(native),
+    D => Key::Unidentified(native),
+    E => Key::Unidentified(native),
+    F => Key::Unidentified(native),
+    G => Key::Unidentified(native),
+    H => Key::Unidentified(native),
+    I => Key::Unidentified(native),
+    J => Key::Unidentified(native),
+    K => Key::Unidentified(native),
+    L => Key::Unidentified(native),
+    M => Key::Unidentified(native),
+    N => Key::Unidentified(native),
+    O => Key::Unidentified(native),
+    P => Key::Unidentified(native),
+    Q => Key::Unidentified(native),
+    R => Key::Unidentified(native),
+    S => Key::Unidentified(native),
+    T => Key::Unidentified(native),
+    U => Key::Unidentified(native),
+    V => Key::Unidentified(native),
+    W => Key::Unidentified(native),
+    X => Key::Unidentified(native),
+    Y => Key::Unidentified(native),
+    Z => Key::Unidentified(native),
+    Comma => Key::Unidentified(native),
+    Period => Key::Unidentified(native),
+    Grave => Key::Unidentified(native),
+    Minus => Key::Unidentified(native),
+    Equals => Key::Unidentified(native),
+    LeftBracket => Key::Unidentified(native),
+    RightBracket => Key::Unidentified(native),
+    Backslash => Key::Unidentified(native),
+    Semicolon => Key::Unidentified(native),
+    Apostrophe => Key::Unidentified(native),
+    Slash => Key::Unidentified(native),
+    At => Key::Unidentified(native),
+    Plus => Key::Unidentified(native),
+    //-------------------------------------------------------------------------------
+    DpadUp => Key::ArrowUp,
+    DpadDown => Key::ArrowDown,
+    DpadLeft => Key::ArrowLeft,
+    DpadRight => Key::ArrowRight,
+    DpadCenter => Key::Enter,
+
+    VolumeUp => Key::AudioVolumeUp,
+    VolumeDown => Key::AudioVolumeDown,
+    Power => Key::Power,
+    Camera => Key::Camera,
+    // Clear => Key::Named(NamedKey::Clear),
+    AltLeft => Key::Alt,
+    AltRight => Key::Alt,
+    ShiftLeft => Key::Shift,
+    ShiftRight => Key::Shift,
+    Tab => Key::Tab,
+    Space => Key::Space,
+    Sym => Key::Symbol,
+    Explorer => Key::LaunchWebBrowser,
+    Envelope => Key::LaunchMail,
+    Enter => Key::Enter,
+    Del => Key::Backspace,
+
+    // According to https://developer.android.com/reference/android/view/KeyEvent#KEYCODE_NUM
+    // Num => Key::Named(NamedKey::Alt),
+
+    // Headsethook => Key::Named(NamedKey::HeadsetHook),
+    // Focus => Key::Named(NamedKey::CameraFocus),
+
+    // Notification => Key::Named(NamedKey::Notification),
+    // Search => Key::Named(NamedKey::BrowserSearch),
+    MediaPlayPause => Key::MediaPlayPause,
+    MediaStop => Key::MediaStop,
+    MediaNext => Key::MediaTrackNext,
+    MediaPrevious => Key::MediaTrackPrevious,
+    MediaRewind => Key::MediaRewind,
+    MediaFastForward => Key::MediaFastForward,
+    Mute => Key::MicrophoneVolumeMute,
+    PageUp => Key::PageUp,
+    PageDown => Key::PageDown,
+
+    Escape => Key::Escape,
+    ForwardDel => Key::Delete,
+    CtrlLeft => Key::Control,
+    CtrlRight => Key::Control,
+    CapsLock => Key::CapsLock,
+    ScrollLock => Key::ScrollLock,
+    MetaLeft => Key::Super,
+    MetaRight => Key::Super,
+    Function => Key::Fn,
+    SysRq => Key::PrintScreen,
+    Break => Key::Pause,
+    MoveHome => Key::Home,
+    MoveEnd => Key::End,
+    Insert => Key::Insert,
+    Forward => Key::BrowserForward,
+    MediaPlay => Key::MediaPlay,
+    MediaPause => Key::MediaPause,
+    MediaClose => Key::MediaClose,
+    MediaEject => Key::Eject,
+    MediaRecord => Key::MediaRecord,
+    F1 => Key::F1,
+    F2 => Key::F2,
+    F3 => Key::F3,
+    F4 => Key::F4,
+    F5 => Key::F5,
+    F6 => Key::F6,
+    F7 => Key::F7,
+    F8 => Key::F8,
+    F9 => Key::F9,
+    F10 => Key::F10,
+    F11 => Key::F11,
+    F12 => Key::F12,
+    NumLock => Key::NumLock,
+    Numpad0 => Key::Unidentified(native),
+    Numpad1 => Key::Unidentified(native),
+    Numpad2 => Key::Unidentified(native),
+    Numpad3 => Key::Unidentified(native),
+    Numpad4 => Key::Unidentified(native),
+    Numpad5 => Key::Unidentified(native),
+    Numpad6 => Key::Unidentified(native),
+    Numpad7 => Key::Unidentified(native),
+    Numpad8 => Key::Unidentified(native),
+    Numpad9 => Key::Unidentified(native),
+    NumpadDivide => Key::Unidentified(native),
+    NumpadMultiply => Key::Unidentified(native),
+    NumpadSubtract => Key::Unidentified(native),
+    NumpadAdd => Key::Unidentified(native),
+    NumpadDot => Key::Unidentified(native),
+    NumpadComma => Key::Unidentified(native),
+    NumpadEnter => Key::Unidentified(native),
+    NumpadEquals => Key::Unidentified(native),
+    NumpadLeftParen => Key::Unidentified(native),
+    NumpadRightParen => Key::Unidentified(native),
+
+    VolumeMute => Key::AudioVolumeMute,
+    Info => Key::Info,
+    ChannelUp => Key::ChannelUp,
+    ChannelDown => Key::ChannelDown,
+    ZoomIn => Key::ZoomIn,
+    ZoomOut => Key::ZoomOut,
+    TV => Key::TV,
+    // Guide => Key::Named(NamedKey::Guide),
+    // Dvr => Key::Named(NamedKey::DVR),
+    // Bookmark => Key::Named(NamedKey::BrowserFavorites),
+    // Captions => Key::Named(NamedKey::ClosedCaptionToggle),
+    // Settings => Key::Named(NamedKey::Settings),
+    // TvPower => Key::Named(NamedKey::TVPower),
+    // TvInput => Key::Named(NamedKey::TVInput),
+    // StbPower => Key::Named(NamedKey::STBPower),
+    // StbInput => Key::Named(NamedKey::STBInput),
+    // AvrPower => Key::Named(NamedKey::AVRPower),
+    // AvrInput => Key::Named(NamedKey::AVRInput),
+    // ProgRed => Key::Named(NamedKey::ColorF0Red),
+    // ProgGreen => Key::Named(NamedKey::ColorF1Green),
+    // ProgYellow => Key::Named(NamedKey::ColorF2Yellow),
+    // ProgBlue => Key::Named(NamedKey::ColorF3Blue),
+    // AppSwitch => Key::Named(NamedKey::AppSwitch),
+    // LanguageSwitch => Key::Named(NamedKey::GroupNext),
+    // MannerMode => Key::Named(NamedKey::MannerMode),
+    // Keycode3dMode => Key::Named(NamedKey::TV3DMode),
+    // Contacts => Key::Named(NamedKey::LaunchContacts),
+    Calendar => Key::LaunchCalendar,
+    // Music => Key::Named(NamedKey::LaunchMusicPlayer),
+    // Calculator => Key::Named(NamedKey::LaunchApplication2),
+    ZenkakuHankaku => Key::ZenkakuHankaku,
+    // Eisu => Key::Named(NamedKey::Eisu),
+    Muhenkan => Key::NonConvert,
+    Henkan => Key::Convert,
+    KatakanaHiragana => Key::HiraganaKatakana,
+    // Kana => Key::Named(NamedKey::KanjiMode),
+    BrightnessDown => Key::BrightnessDown,
+    BrightnessUp => Key::BrightnessUp,
+    // MediaAudioTrack => Key::Named(NamedKey::MediaAudioTrack),
+    Sleep => Key::Standby,
+    Wakeup => Key::WakeUp,
+    // Pairing => Key::Named(NamedKey::Pairing),
+    // MediaTopMenu => Key::Named(NamedKey::MediaTopMenu),
+    // LastChannel => Key::Named(NamedKey::MediaLast),
+    // TvDataService => Key::Named(NamedKey::TVDataService),
+    // VoiceAssist => Key::Named(NamedKey::VoiceDial),
+    // TvRadioService => Key::Named(NamedKey::TVRadioService),
+    // TvTeletext => Key::Named(NamedKey::Teletext),
+    // TvNumberEntry => Key::Named(NamedKey::TVNumberEntry),
+    // TvTerrestrialAnalog => Key::Named(NamedKey::TVTerrestrialAnalog),
+    // TvTerrestrialDigital => Key::Named(NamedKey::TVTerrestrialDigital),
+    // TvSatellite => Key::Named(NamedKey::TVSatellite),
+    // TvSatelliteBs => Key::Named(NamedKey::TVSatelliteBS),
+    // TvSatelliteCs => Key::Named(NamedKey::TVSatelliteCS),
+    // TvSatelliteService => Key::Named(NamedKey::TVSatelliteToggle),
+    // TvNetwork => Key::Named(NamedKey::TVNetwork),
+    // TvAntennaCable => Key::Named(NamedKey::TVAntennaCable),
+    // TvInputHdmi1 => Key::Named(NamedKey::TVInputHDMI1),
+    // TvInputHdmi2 => Key::Named(NamedKey::TVInputHDMI2),
+    // TvInputHdmi3 => Key::Named(NamedKey::TVInputHDMI3),
+    // TvInputHdmi4 => Key::Named(NamedKey::TVInputHDMI4),
+    // TvInputComposite1 => Key::Named(NamedKey::TVInputComposite1),
+    // TvInputComposite2 => Key::Named(NamedKey::TVInputComposite2),
+    // TvInputComponent1 => Key::Named(NamedKey::TVInputComponent1),
+    // TvInputComponent2 => Key::Named(NamedKey::TVInputComponent2),
+    // TvInputVga1 => Key::Named(NamedKey::TVInputVGA1),
+    // TvAudioDescription => Key::Named(NamedKey::TVAudioDescription),
+    // TvAudioDescriptionMixUp => Key::Named(NamedKey::TVAudioDescriptionMixUp),
+    // TvAudioDescriptionMixDown => Key::Named(NamedKey::TVAudioDescriptionMixDown),
+    // TvZoomMode => Key::Named(NamedKey::ZoomToggle),
+    // TvContentsMenu => Key::Named(NamedKey::TVContentsMenu),
+    // TvMediaContextMenu => Key::Named(NamedKey::TVMediaContext),
+    // TvTimerProgramming => Key::Named(NamedKey::TVTimer),
+    Help => Key::Help,
+    // NavigatePrevious => Key::Named(NamedKey::NavigatePrevious),
+    // NavigateNext => Key::Named(NamedKey::NavigateNext),
+    // NavigateIn => Key::Named(NamedKey::NavigateIn),
+    // NavigateOut => Key::Named(NamedKey::NavigateOut),
+    // MediaSkipForward => Key::Named(NamedKey::MediaSkipForward),
+    // MediaSkipBackward => Key::Named(NamedKey::MediaSkipBackward),
+    // MediaStepForward => Key::Named(NamedKey::MediaStepForward),
+    // MediaStepBackward => Key::Named(NamedKey::MediaStepBackward),
+    Cut => Key::Cut,
+    Copy => Key::Copy,
+    Paste => Key::Paste,
+    Refresh => Key::BrowserRefresh,
+
+    // -----------------------------------------------------------------
+    // Keycodes that don't have a logical Key mapping
+    // -----------------------------------------------------------------
+    Unknown => Key::Unidentified(native),
+
+    // Can be added on demand
+    // SoftLeft => Key::Unidentified(native),
+    // SoftRight => Key::Unidentified(native),
+    Menu => Key::Unidentified(native),
+
+    // Pictsymbols => Key::Unidentified(native),
+    // SwitchCharset => Key::Unidentified(native),
+
+    // -----------------------------------------------------------------
+    // Gamepad events should be exposed through a separate API, not
+    // keyboard events
+    // ButtonA => Key::Unidentified(native),
+    // ButtonB => Key::Unidentified(native),
+    // ButtonC => Key::Unidentified(native),
+    // ButtonX => Key::Unidentified(native),
+    // ButtonY => Key::Unidentified(native),
+    // ButtonZ => Key::Unidentified(native),
+    // ButtonL1 => Key::Unidentified(native),
+    // ButtonR1 => Key::Unidentified(native),
+    // ButtonL2 => Key::Unidentified(native),
+    // ButtonR2 => Key::Unidentified(native),
+    // ButtonThumbl => Key::Unidentified(native),
+    // ButtonThumbr => Key::Unidentified(native),
+    // ButtonStart => Key::Unidentified(native),
+    // ButtonSelect => Key::Unidentified(native),
+    // ButtonMode => Key::Unidentified(native),
+    // // -----------------------------------------------------------------
+    // Window => Key::Unidentified(native),
+
+    // Button1 => Key::Unidentified(native),
+    // Button2 => Key::Unidentified(native),
+    // Button3 => Key::Unidentified(native),
+    // Button4 => Key::Unidentified(native),
+    // Button5 => Key::Unidentified(native),
+    // Button6 => Key::Unidentified(native),
+    // Button7 => Key::Unidentified(native),
+    // Button8 => Key::Unidentified(native),
+    // Button9 => Key::Unidentified(native),
+    // Button10 => Key::Unidentified(native),
+    // Button11 => Key::Unidentified(native),
+    // Button12 => Key::Unidentified(native),
+    // Button13 => Key::Unidentified(native),
+    // Button14 => Key::Unidentified(native),
+    // Button15 => Key::Unidentified(native),
+    // Button16 => Key::Unidentified(native),
+    Yen => Key::Unidentified(native),
+    Ro => Key::Unidentified(native),
+
+    // Assist => Key::Unidentified(native),
+
+    // Keycode11 => Key::Unidentified(native),
+    // Keycode12 => Key::Unidentified(native),
+
+    // StemPrimary => Key::Unidentified(native),
+    // Stem1 => Key::Unidentified(native),
+    // Stem2 => Key::Unidentified(native),
+    // Stem3 => Key::Unidentified(native),
+
+    // DpadUpLeft => Key::Unidentified(native),
+    // DpadDownLeft => Key::Unidentified(native),
+    // DpadUpRight => Key::Unidentified(native),
+    // DpadDownRight => Key::Unidentified(native),
+
+    // SoftSleep => Key::Unidentified(native),
+
+    // SystemNavigationUp => Key::Unidentified(native),
+    // SystemNavigationDown => Key::Unidentified(native),
+    // SystemNavigationLeft => Key::Unidentified(native),
+    // SystemNavigationRight => Key::Unidentified(native),
+
+    // AllApps => Key::Unidentified(native),
+    // ThumbsUp => Key::Unidentified(native),
+    // ThumbsDown => Key::Unidentified(native),
+    // ProfileSwitch => Key::Unidentified(native),
+
+    // It's always possible that new versions of Android could introduce
+    // key codes we can't know about at compile time.
+    _ => Key::Unidentified(native),
+  }
+}
+
+pub fn to_location(keycode: Keycode) -> KeyLocation {
+  use openharmony_ability::xcomponent::KeyCode::*;
+
+  match keycode {
+    AltLeft => KeyLocation::Left,
+    AltRight => KeyLocation::Right,
+    ShiftLeft => KeyLocation::Left,
+    ShiftRight => KeyLocation::Right,
+
+    CtrlLeft => KeyLocation::Left,
+    CtrlRight => KeyLocation::Right,
+    MetaLeft => KeyLocation::Left,
+    MetaRight => KeyLocation::Right,
+
+    NumLock => KeyLocation::Numpad,
+    Numpad0 => KeyLocation::Numpad,
+    Numpad1 => KeyLocation::Numpad,
+    Numpad2 => KeyLocation::Numpad,
+    Numpad3 => KeyLocation::Numpad,
+    Numpad4 => KeyLocation::Numpad,
+    Numpad5 => KeyLocation::Numpad,
+    Numpad6 => KeyLocation::Numpad,
+    Numpad7 => KeyLocation::Numpad,
+    Numpad8 => KeyLocation::Numpad,
+    Numpad9 => KeyLocation::Numpad,
+    NumpadDivide => KeyLocation::Numpad,
+    NumpadMultiply => KeyLocation::Numpad,
+    NumpadSubtract => KeyLocation::Numpad,
+    NumpadAdd => KeyLocation::Numpad,
+    NumpadDot => KeyLocation::Numpad,
+    NumpadComma => KeyLocation::Numpad,
+    NumpadEnter => KeyLocation::Numpad,
+    NumpadEquals => KeyLocation::Numpad,
+    NumpadLeftParen => KeyLocation::Numpad,
+    NumpadRightParen => KeyLocation::Numpad,
+
+    _ => KeyLocation::Standard,
+  }
+}
diff --git a/src/platform_impl/ohos/mod.rs b/src/platform_impl/ohos/mod.rs
new file mode 100644
index 0000000..b2c93ce
--- /dev/null
+++ b/src/platform_impl/ohos/mod.rs
@@ -0,0 +1,904 @@
+use std::cell::{Cell, RefCell};
+use std::collections::VecDeque;
+use std::hash::Hash;
+use std::marker::PhantomData;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::mpsc;
+
+use keycodes::{to_location, to_logical};
+use openharmony_ability::xcomponent::{Action, TouchEvent};
+
+use openharmony_ability::{
+  ime::KeyboardStatus, Configuration, Event as MainEvent, ImeEvent, InputEvent, OpenHarmonyApp,
+  OpenHarmonyWaker, Rect,
+};
+
+use crate::dpi::{PhysicalPosition, PhysicalSize, Position, Size};
+use crate::error::{self};
+use crate::event::{self, ElementState, Force, StartCause};
+use crate::event_loop::{self, ControlFlow};
+use crate::keyboard::{Key, KeyCode, KeyLocation, NativeKeyCode};
+use crate::monitor;
+use crate::window::{self, Fullscreen, ResizeDirection, Theme, WindowSizeConstraints};
+
+mod keycodes;
+
+pub(crate) use crate::icon::NoIcon as PlatformIcon;
+
+static HAS_FOCUS: AtomicBool = AtomicBool::new(true);
+
+struct PeekableReceiver<T> {
+  recv: mpsc::Receiver<T>,
+  first: Option<T>,
+}
+
+impl<T> PeekableReceiver<T> {
+  pub fn from_recv(recv: mpsc::Receiver<T>) -> Self {
+    Self { recv, first: None }
+  }
+
+  pub fn try_recv(&mut self) -> Result<T, mpsc::TryRecvError> {
+    if let Some(first) = self.first.take() {
+      return Ok(first);
+    }
+    self.recv.try_recv()
+  }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct KeyEventExtra {}
+
+pub struct EventLoop<T: 'static> {
+  pub(crate) openharmony_app: OpenHarmonyApp,
+  window_target: event_loop::EventLoopWindowTarget<T>,
+  cause: StartCause,
+  user_events_sender: mpsc::Sender<T>,
+  user_events_receiver: PeekableReceiver<T>,
+  event_loop: RefCell<Option<Box<dyn FnMut(event::Event<T>)>>>,
+}
+
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub(crate) struct PlatformSpecificEventLoopAttributes {
+  pub(crate) openharmony_app: Option<OpenHarmonyApp>,
+}
+
+impl Default for PlatformSpecificEventLoopAttributes {
+  fn default() -> Self {
+    Self {
+      openharmony_app: Default::default(),
+    }
+  }
+}
+
+impl<T: 'static> EventLoop<T> {
+  pub(crate) fn new(attributes: &PlatformSpecificEventLoopAttributes) -> Self {
+    let (user_events_sender, user_events_receiver) = mpsc::channel();
+
+    let openharmony_app = attributes.openharmony_app.as_ref().expect(
+      "An `OpenHarmonyApp` as passed to lib is required to create an `EventLoop` on \
+             OpenHarmony or HarmonyNext",
+    );
+
+    Self {
+      openharmony_app: openharmony_app.clone(),
+      window_target: event_loop::EventLoopWindowTarget {
+        p: EventLoopWindowTarget {
+          app: openharmony_app.clone(),
+          control_flow: Cell::new(ControlFlow::default()),
+          exit: Cell::new(false),
+          _marker: PhantomData,
+        },
+        _marker: PhantomData,
+      },
+      cause: StartCause::Init,
+      user_events_sender,
+      user_events_receiver: PeekableReceiver::from_recv(user_events_receiver),
+      event_loop: RefCell::new(None),
+    }
+  }
+
+  pub(crate) fn window_target(&self) -> &event_loop::EventLoopWindowTarget<T> {
+    &self.window_target
+  }
+
+  // TODO: For input event, we need some real examples to test it
+  fn handle_input_event(&self, event: &InputEvent) {
+    match event {
+      InputEvent::TouchEvent(motion_event) => {
+        let window_id = window::WindowId(WindowId);
+        let device_id = event::DeviceId(DeviceId(motion_event.device_id as _));
+        let action = motion_event.event_type;
+
+        let phase = match motion_event.event_type {
+          TouchEvent::Down => Some(event::TouchPhase::Started),
+          TouchEvent::Up => Some(event::TouchPhase::Ended),
+          TouchEvent::Move => Some(event::TouchPhase::Moved),
+          TouchEvent::Cancel => Some(event::TouchPhase::Cancelled),
+          _ => {
+            None // TODO mouse events
+          }
+        };
+
+        if let Some(phase) = phase {
+          for pointer in motion_event.touch_points.iter() {
+            let position = PhysicalPosition {
+              x: pointer.x as _,
+              y: pointer.y as _,
+            };
+            trace!(
+              "Input event {device_id:?}, {action:?}, loc={position:?}, \
+                                 pointer={pointer:?}"
+            );
+
+            let event = event::Event::WindowEvent {
+              window_id,
+              event: event::WindowEvent::Touch(event::Touch {
+                device_id,
+                phase,
+                location: position,
+                id: pointer.id as u64,
+                force: Some(Force::Normalized(pointer.force as f64)),
+              }),
+            };
+            if let Some(ref mut h) = *self.event_loop.borrow_mut() {
+              h(event);
+            }
+          }
+        }
+      }
+      InputEvent::KeyEvent(key) => {
+        match key.code {
+          keycode => {
+            let state = match key.action {
+              Action::Down => event::ElementState::Pressed,
+              Action::Up => event::ElementState::Released,
+              _ => event::ElementState::Released,
+            };
+
+            let native = NativeKeyCode::Ohos(keycode.into());
+            let physical_key = KeyCode::Unidentified(native);
+            let logical_key = to_logical(keycode);
+
+            let event = event::Event::WindowEvent {
+              window_id: window::WindowId(WindowId),
+              event: event::WindowEvent::KeyboardInput {
+                device_id: event::DeviceId(DeviceId(key.device_id as _)),
+                event: event::KeyEvent {
+                  state,
+                  physical_key,
+                  logical_key,
+                  location: to_location(keycode),
+                  // TODO
+                  repeat: false,
+                  text: None,
+                  platform_specific: KeyEventExtra {},
+                },
+                is_synthetic: false,
+              },
+            };
+            if let Some(ref mut h) = *self.event_loop.borrow_mut() {
+              h(event);
+            }
+          }
+        }
+      }
+      InputEvent::ImeEvent(data) => match data {
+        ImeEvent::TextInputEvent(s) => {
+          if let Some(ref mut h) = *self.event_loop.borrow_mut() {
+            h(event::Event::WindowEvent {
+              window_id: window::WindowId(WindowId),
+              event: event::WindowEvent::ReceivedImeText(s.text.clone()),
+            })
+          }
+        }
+        ImeEvent::BackspaceEvent(_) => {
+          if let Some(ref mut h) = *self.event_loop.borrow_mut() {
+            // Mock keyboard input event
+            let _ = [ElementState::Pressed, ElementState::Released].map(|state| {
+              h(event::Event::WindowEvent {
+                window_id: window::WindowId(WindowId),
+                event: event::WindowEvent::KeyboardInput {
+                  device_id: event::DeviceId(DeviceId(0)),
+                  event: event::KeyEvent {
+                    state,
+                    logical_key: Key::Backspace,
+                    physical_key: KeyCode::Backspace,
+                    platform_specific: KeyEventExtra {},
+                    repeat: false,
+                    location: KeyLocation::Standard,
+                    text: None,
+                  },
+                  is_synthetic: false,
+                },
+              });
+            });
+          }
+        }
+        ImeEvent::EnterEvent(_) => {
+          if let Some(ref mut h) = *self.event_loop.borrow_mut() {
+            // Mock keyboard input event
+            // Mock keyboard input event
+            let _ = [ElementState::Pressed, ElementState::Released].map(|state| {
+              h(event::Event::WindowEvent {
+                window_id: window::WindowId(WindowId),
+                event: event::WindowEvent::KeyboardInput {
+                  device_id: event::DeviceId(DeviceId(0)),
+                  event: event::KeyEvent {
+                    state,
+                    logical_key: Key::Enter,
+                    physical_key: KeyCode::Enter,
+                    platform_specific: KeyEventExtra {},
+                    repeat: false,
+                    location: KeyLocation::Standard,
+                    text: None,
+                  },
+                  is_synthetic: false,
+                },
+              });
+            });
+          }
+        }
+        ImeEvent::ImeStatusEvent(s) => match s {
+          KeyboardStatus::Hide => {
+            if let Some(ref mut h) = *self.event_loop.borrow_mut() {
+              // Mock keyboard input event that make sure egui can receive the event and trigger onblur event
+              [ElementState::Pressed, ElementState::Released].map(|state| {
+                h(event::Event::WindowEvent {
+                  window_id: window::WindowId(WindowId),
+                  event: event::WindowEvent::KeyboardInput {
+                    device_id: event::DeviceId(DeviceId(0)),
+                    event: event::KeyEvent {
+                      state,
+                      logical_key: Key::Enter,
+                      physical_key: KeyCode::Enter,
+                      platform_specific: KeyEventExtra {},
+                      repeat: false,
+                      location: KeyLocation::Standard,
+                      text: None,
+                    },
+                    is_synthetic: false,
+                  },
+                });
+              });
+            }
+          }
+          _ => {
+            warn!("Unknown openharmony_ability ime status event {s:?}")
+          }
+        },
+      },
+      _ => {
+        warn!("Unknown openharmony_ability input event {event:?}")
+      }
+    }
+  }
+
+  pub fn run<F>(self, event_handler: F) -> ()
+  where
+    F: FnMut(event::Event<T>, &event_loop::EventLoopWindowTarget<T>, &mut ControlFlow),
+  {
+    let event_looper = Box::leak(Box::new(self));
+    event_looper.run_return(event_handler);
+  }
+
+  pub fn run_return<F>(&mut self, mut event_handle: F) -> i32
+  where
+    F: FnMut(event::Event<T>, &event_loop::EventLoopWindowTarget<T>, &mut ControlFlow),
+  {
+    let mut control_flow = ControlFlow::default();
+    let target = &self.window_target;
+
+    {
+      let handle = unsafe {
+        std::mem::transmute::<Box<dyn FnMut(event::Event<T>)>, Box<dyn FnMut(event::Event<T>)>>(
+          Box::new(move |e| {
+            event_handle(e, &target, &mut control_flow);
+            // We need to dispatch it after every event callbacks.
+            event_handle(event::Event::MainEventsCleared, &target, &mut control_flow);
+          }),
+        )
+      };
+      self.event_loop.replace(Some(handle));
+    }
+
+    self.openharmony_app.clone().run_loop(|event| {
+      match event {
+        MainEvent::SurfaceCreate { .. } => {
+          if let Some(ref mut h) = *self.event_loop.borrow_mut() {
+            h(event::Event::NewEvents(StartCause::Init));
+            h(event::Event::WindowEvent { window_id: window::WindowId(WindowId), event: event::WindowEvent::Resumed });
+          }
+        }
+        MainEvent::SurfaceDestroy { .. } => {
+          if let Some(ref mut h) = *self.event_loop.borrow_mut() {
+            h(event::Event::WindowEvent { window_id: window::WindowId(WindowId), event: event::WindowEvent::Suspended });
+          }
+        }
+        MainEvent::WindowResize(size) => {
+          let size = PhysicalSize::new(size.width as _, size.height as _);
+          let event = event::Event::WindowEvent {
+            window_id: window::WindowId(WindowId),
+            event: event::WindowEvent::Resized(size),
+          };
+
+          if let Some(ref mut h) = *self.event_loop.borrow_mut() {
+            h(event);
+          }
+        }
+        MainEvent::WindowRedraw { .. } => {
+          let event = event::Event::RedrawRequested(window::WindowId(WindowId));
+
+          if let Some(ref mut h) = *self.event_loop.borrow_mut() {
+            h(event);
+          }
+        }
+        MainEvent::ContentRectChange { .. } => {
+          warn!("TODO: find a way to notify application of content rect change");
+        }
+        MainEvent::GainedFocus => {
+          HAS_FOCUS.store(true, Ordering::Relaxed);
+
+          if let Some(ref mut h) = *self.event_loop.borrow_mut() {
+            h(event::Event::WindowEvent {
+              window_id: window::WindowId(WindowId),
+              event: event::WindowEvent::Focused(true),
+            });
+          }
+        }
+        MainEvent::LostFocus => {
+          HAS_FOCUS.store(false, Ordering::Relaxed);
+
+          if let Some(ref mut h) = *self.event_loop.borrow_mut() {
+            h(event::Event::WindowEvent {
+              window_id: window::WindowId(WindowId),
+              event: event::WindowEvent::Focused(true),
+            });
+          }
+        }
+        MainEvent::ConfigChanged { .. } => {
+          let size = self.openharmony_app.content_rect();
+          let scale = self.openharmony_app.scale();
+          let mut size = PhysicalSize::new(size.width as _, size.height as _);
+          let event = event::Event::WindowEvent {
+            window_id: window::WindowId(WindowId),
+            event: event::WindowEvent::ScaleFactorChanged {
+              new_inner_size: &mut size,
+              scale_factor: scale as _,
+            },
+          };
+
+          if let Some(ref mut h) = *self.event_loop.borrow_mut() {
+            h(event);
+          }
+        }
+        MainEvent::Start => {
+          // XXX: how to forward this state to applications?
+          warn!("TODO: forward onStart notification to application");
+        }
+        MainEvent::Resume { .. } => {
+          if let Some(ref mut h) = *self.event_loop.borrow_mut() {
+            h(event::Event::WindowEvent { window_id: window::WindowId(WindowId), event: event::WindowEvent::Resumed });
+          }
+        }
+        MainEvent::SaveState { .. } => {
+          // XXX: how to forward this state to applications?
+          // XXX: also how do we expose state restoration to apps?
+          warn!("TODO: forward saveState notification to application");
+        }
+        MainEvent::Pause => {
+          debug!("App Paused - stopped running");
+          // TODO: This is incorrect - will be solved in https://github.com/rust-windowing/winit/pull/3897
+          // self.running = false;
+        }
+        MainEvent::WindowDestroy => {
+          if let Some(ref mut h) = *self.event_loop.borrow_mut() {
+            let e = event::Event::WindowEvent {
+              window_id: window::WindowId(WindowId),
+              event: event::WindowEvent::CloseRequested,
+            };
+            h(e);
+          }
+        }
+        MainEvent::Destroy => {
+          // XXX: maybe exit mainloop to drop things before being
+          // killed by the OS?
+          warn!("TODO: forward onDestroy notification to application");
+        }
+        MainEvent::Input(input_event) => {
+          self.handle_input_event(&input_event);
+        }
+        MainEvent::UserEvent { .. } => {
+          if let Some(ref mut h) = *self.event_loop.borrow_mut() {
+            if let Ok(event) = self.user_events_receiver.try_recv() {
+              let event = event::Event::UserEvent(event);
+              h(event);
+            }
+          }
+        }
+        unknown => {
+          trace!("Unknown MainEvent {unknown:?} (ignored)");
+        }
+      };
+
+      if self.window_target.p.exit.get() {
+        if let Some(ref mut h) = *self.event_loop.borrow_mut() {
+          h(event::Event::LoopDestroyed);
+          self.openharmony_app.exit(0);
+        }
+      }
+    });
+    0
+  }
+
+  pub fn create_proxy(&self) -> EventLoopProxy<T> {
+    EventLoopProxy {
+      user_events_sender: self.user_events_sender.clone(),
+      waker: self.openharmony_app.create_waker(),
+    }
+  }
+}
+
+pub struct EventLoopProxy<T: 'static> {
+  user_events_sender: mpsc::Sender<T>,
+  waker: OpenHarmonyWaker,
+}
+
+impl<T: 'static> EventLoopProxy<T> {
+  pub fn send_event(&self, event: T) -> Result<(), event_loop::EventLoopClosed<T>> {
+    self
+      .user_events_sender
+      .send(event)
+      .map_err(|err| event_loop::EventLoopClosed(err.0))?;
+    self.waker.wake();
+    Ok(())
+  }
+}
+
+impl<T: 'static> Clone for EventLoopProxy<T> {
+  fn clone(&self) -> Self {
+    EventLoopProxy {
+      user_events_sender: self.user_events_sender.clone(),
+      waker: self.waker.clone(),
+    }
+  }
+}
+
+#[derive(Clone)]
+pub struct EventLoopWindowTarget<T: 'static> {
+  pub(crate) app: OpenHarmonyApp,
+  control_flow: Cell<ControlFlow>,
+  exit: Cell<bool>,
+  _marker: std::marker::PhantomData<T>,
+}
+
+impl<T: 'static> EventLoopWindowTarget<T> {
+  pub fn available_monitors(&self) -> VecDeque<MonitorHandle> {
+    let mut v = VecDeque::with_capacity(1);
+    v.push_back(MonitorHandle::new(self.app.clone()));
+    v
+  }
+
+  pub fn primary_monitor(&self) -> Option<monitor::MonitorHandle> {
+    Some(monitor::MonitorHandle {
+      inner: MonitorHandle::new(self.app.clone()),
+    })
+  }
+
+  #[inline]
+  pub fn monitor_from_point(&self, _x: f64, _y: f64) -> Option<MonitorHandle> {
+    warn!("`Window::monitor_from_point` is ignored on OpenHarmony");
+    return None;
+  }
+
+  #[cfg(feature = "rwh_05")]
+  #[inline]
+  pub fn raw_display_handle_rwh_05(&self) -> rwh_05::RawDisplayHandle {
+    unreachable!("rwh_05 is not supported on OpenHarmony");
+  }
+
+  #[cfg(feature = "rwh_06")]
+  #[inline]
+  pub fn raw_display_handle_rwh_06(&self) -> Result<rwh_06::RawDisplayHandle, rwh_06::HandleError> {
+    Ok(rwh_06::RawDisplayHandle::Ohos(
+      rwh_06::OhosDisplayHandle::new(),
+    ))
+  }
+
+  pub fn cursor_position(&self) -> Result<PhysicalPosition<f64>, error::ExternalError> {
+    debug!("`EventLoopWindowTarget::cursor_position` is ignored on OpenHarmony");
+    Ok((0, 0).into())
+  }
+}
+
+#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub(crate) struct WindowId;
+
+impl WindowId {
+  pub const fn dummy() -> Self {
+    WindowId
+  }
+}
+
+impl From<WindowId> for u64 {
+  fn from(_: WindowId) -> Self {
+    0
+  }
+}
+
+impl From<u64> for WindowId {
+  fn from(_: u64) -> Self {
+    Self
+  }
+}
+
+#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct DeviceId(i32);
+
+impl DeviceId {
+  pub const fn dummy() -> Self {
+    DeviceId(0)
+  }
+}
+
+#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
+pub struct PlatformSpecificWindowBuilderAttributes;
+
+pub(crate) struct Window {
+  app: OpenHarmonyApp,
+}
+
+impl Window {
+  pub(crate) fn new<T: 'static>(
+    el: &EventLoopWindowTarget<T>,
+    _window_attrs: window::WindowAttributes,
+    _: PlatformSpecificWindowBuilderAttributes,
+  ) -> Result<Self, error::OsError> {
+    // FIXME this ignores requested window attributes
+
+    Ok(Self {
+      app: el.app.clone(),
+    })
+  }
+
+  pub fn request_redraw(&self) {}
+
+  #[inline]
+  pub fn monitor_from_point(&self, _x: f64, _y: f64) -> Option<monitor::MonitorHandle> {
+    warn!("`Window::monitor_from_point` is ignored on OpenHarmony");
+    return None;
+  }
+
+  pub fn id(&self) -> WindowId {
+    WindowId
+  }
+
+  pub fn scale_factor(&self) -> f64 {
+    self.app.scale() as f64
+  }
+
+  pub fn available_monitors(&self) -> VecDeque<MonitorHandle> {
+    let mut v = VecDeque::with_capacity(1);
+    v.push_back(MonitorHandle::new(self.app.clone()));
+    v
+  }
+
+  pub fn inner_position(&self) -> Result<PhysicalPosition<i32>, error::NotSupportedError> {
+    Err(error::NotSupportedError::new())
+  }
+
+  pub fn inner_size(&self) -> PhysicalSize<u32> {
+    self.outer_size()
+  }
+
+  pub fn set_inner_size(&self, _size: Size) {
+    warn!("Cannot set window size on OpenHarmony");
+  }
+  pub fn set_inner_size_constraints(&self, _: WindowSizeConstraints) {}
+
+  pub fn outer_position(&self) -> Result<PhysicalPosition<i32>, error::NotSupportedError> {
+    Err(error::NotSupportedError::new())
+  }
+
+  pub fn set_outer_position(&self, _position: Position) {
+    // no effect
+  }
+
+  pub fn outer_size(&self) -> PhysicalSize<u32> {
+    MonitorHandle::new(self.app.clone()).size()
+  }
+
+  pub fn set_min_inner_size(&self, _: Option<Size>) {}
+
+  pub fn set_max_inner_size(&self, _: Option<Size>) {}
+
+  pub fn set_title(&self, _title: &str) {}
+
+  pub fn set_visible(&self, _visibility: bool) {}
+
+  pub fn set_focus(&self) {
+    //FIXME: implementation goes here
+    warn!("set_focus not yet implemented on OpenHarmony");
+  }
+
+  pub fn set_focusable(&self, _focusable: bool) {
+    warn!("set_focusable not yet implemented on OpenHarmony");
+  }
+
+  pub fn is_focused(&self) -> bool {
+    log::warn!("`Window::is_focused` is ignored on OpenHarmony");
+    false
+  }
+
+  pub fn is_always_on_top(&self) -> bool {
+    log::warn!("`Window::is_always_on_top` is ignored on OpenHarmony");
+    false
+  }
+
+  pub fn set_resizable(&self, _resizeable: bool) {
+    warn!("`Window::set_resizable` is ignored on OpenHarmony")
+  }
+
+  pub fn set_minimizable(&self, _minimizable: bool) {
+    warn!("`Window::set_minimizable` is ignored on OpenHarmony")
+  }
+
+  pub fn set_maximizable(&self, _maximizable: bool) {
+    warn!("`Window::set_maximizable` is ignored on OpenHarmony")
+  }
+
+  pub fn set_closable(&self, _closable: bool) {
+    warn!("`Window::set_closable` is ignored on OpenHarmony")
+  }
+
+  pub fn set_minimized(&self, _minimized: bool) {}
+
+  pub fn is_minimized(&self) -> bool {
+    false
+  }
+
+  pub fn set_maximized(&self, _maximized: bool) {}
+
+  pub fn is_maximized(&self) -> bool {
+    false
+  }
+
+  pub fn set_fullscreen(&self, _monitor: Option<Fullscreen>) {
+    warn!("Cannot set fullscreen on OpenHarmony");
+  }
+
+  pub fn fullscreen(&self) -> Option<Fullscreen> {
+    None
+  }
+
+  pub fn set_decorations(&self, _decorations: bool) {}
+  pub fn set_always_on_bottom(&self, _always_on_bottom: bool) {}
+
+  pub fn set_always_on_top(&self, _always_on_top: bool) {}
+  pub fn set_ime_position(&self, _position: Position) {}
+
+  pub fn is_decorated(&self) -> bool {
+    true
+  }
+
+  pub fn is_visible(&self) -> bool {
+    log::warn!("`Window::is_visible` is ignored on OpenHarmony");
+    false
+  }
+
+  pub fn is_resizable(&self) -> bool {
+    warn!("`Window::is_resizable` is ignored on OpenHarmony");
+    false
+  }
+
+  pub fn is_minimizable(&self) -> bool {
+    warn!("`Window::is_minimizable` is ignored on OpenHarmony");
+    false
+  }
+
+  pub fn is_maximizable(&self) -> bool {
+    warn!("`Window::is_maximizable` is ignored on OpenHarmony");
+    false
+  }
+
+  pub fn is_closable(&self) -> bool {
+    warn!("`Window::is_closable` is ignored on OpenHarmony");
+    false
+  }
+
+  pub fn set_window_icon(&self, _window_icon: Option<crate::icon::Icon>) {}
+
+  pub fn set_cursor_icon(&self, _: window::CursorIcon) {}
+  pub fn set_cursor_grab(&self, _: bool) -> Result<(), error::ExternalError> {
+    Err(error::ExternalError::NotSupported(
+      error::NotSupportedError::new(),
+    ))
+  }
+
+  pub fn request_user_attention(&self, _request_type: Option<window::UserAttentionType>) {}
+
+  pub fn set_cursor_position(&self, _: Position) -> Result<(), error::ExternalError> {
+    Err(error::ExternalError::NotSupported(
+      error::NotSupportedError::new(),
+    ))
+  }
+
+  pub fn cursor_position(&self) -> Result<PhysicalPosition<f64>, error::ExternalError> {
+    debug!("`Window::cursor_position` is ignored on OpenHarmony");
+    Ok((0, 0).into())
+  }
+
+  pub fn set_ignore_cursor_events(&self, _ignore: bool) -> Result<(), error::ExternalError> {
+    Err(error::ExternalError::NotSupported(
+      error::NotSupportedError::new(),
+    ))
+  }
+
+  pub fn set_cursor_visible(&self, _: bool) {}
+  pub fn drag_window(&self) -> Result<(), error::ExternalError> {
+    Err(error::ExternalError::NotSupported(
+      error::NotSupportedError::new(),
+    ))
+  }
+
+  pub fn drag_resize_window(
+    &self,
+    _direction: ResizeDirection,
+  ) -> Result<(), error::ExternalError> {
+    Err(error::ExternalError::NotSupported(
+      error::NotSupportedError::new(),
+    ))
+  }
+
+  pub fn set_background_color(&self, _color: Option<crate::window::RGBA>) {}
+
+  pub fn theme(&self) -> Theme {
+    Theme::Light
+  }
+
+  pub fn title(&self) -> String {
+    String::new()
+  }
+
+  #[cfg(feature = "rwh_04")]
+  pub fn raw_window_handle_rwh_04(&self) -> rwh_04::RawWindowHandle {
+    unreachable!("rwh_04 is not supported on OpenHarmony");
+  }
+
+  #[cfg(feature = "rwh_05")]
+  pub fn raw_window_handle_rwh_05(&self) -> rwh_05::RawWindowHandle {
+    unreachable!("rwh_05 is not supported on OpenHarmony");
+  }
+
+  #[cfg(feature = "rwh_05")]
+  pub fn raw_display_handle_rwh_05(&self) -> rwh_05::RawDisplayHandle {
+    unreachable!("rwh_05 is not supported on OpenHarmony");
+  }
+
+  #[cfg(feature = "rwh_06")]
+  // Allow the usage of HasRawWindowHandle inside this function
+  #[allow(deprecated)]
+  pub fn raw_window_handle_rwh_06(&self) -> Result<rwh_06::RawWindowHandle, rwh_06::HandleError> {
+    if let Some(native_window) = self.app.native_window().as_ref() {
+      if let Some(win) = native_window.raw_window_handle() {
+        return Ok(win);
+      }
+      Err(rwh_06::HandleError::Unavailable)
+    } else {
+      Err(rwh_06::HandleError::Unavailable)
+    }
+  }
+
+  #[cfg(feature = "rwh_06")]
+  pub fn raw_display_handle_rwh_06(&self) -> Result<rwh_06::RawDisplayHandle, rwh_06::HandleError> {
+    Ok(rwh_06::RawDisplayHandle::Ohos(
+      rwh_06::OhosDisplayHandle::new(),
+    ))
+  }
+
+  pub fn config(&self) -> Configuration {
+    self.app.config()
+  }
+
+  pub fn content_rect(&self) -> Rect {
+    self.app.content_rect()
+  }
+
+  pub fn current_monitor(&self) -> Option<monitor::MonitorHandle> {
+    Some(monitor::MonitorHandle {
+      inner: MonitorHandle::new(self.app.clone()),
+    })
+  }
+
+  pub fn primary_monitor(&self) -> Option<monitor::MonitorHandle> {
+    Some(monitor::MonitorHandle {
+      inner: MonitorHandle::new(self.app.clone()),
+    })
+  }
+}
+
+#[derive(Default, Clone, Debug)]
+pub struct OsError;
+
+use std::fmt::{self, Display, Formatter};
+impl Display for OsError {
+  fn fmt(&self, fmt: &mut Formatter<'_>) -> Result<(), fmt::Error> {
+    write!(fmt, "OpenHarmony OS Error")
+  }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct MonitorHandle {
+  app: OpenHarmonyApp,
+}
+
+impl MonitorHandle {
+  pub(crate) fn new(app: OpenHarmonyApp) -> Self {
+    Self { app }
+  }
+
+  pub fn name(&self) -> Option<String> {
+    Some("OpenHarmony Device".to_owned())
+  }
+
+  pub fn size(&self) -> PhysicalSize<u32> {
+    let size = self.app.content_rect();
+    PhysicalSize::new(size.width as _, size.height as _)
+  }
+
+  pub fn position(&self) -> PhysicalPosition<i32> {
+    (0, 0).into()
+  }
+
+  pub fn scale_factor(&self) -> f64 {
+    self.app.scale() as f64
+  }
+
+  pub fn video_modes(&self) -> impl Iterator<Item = monitor::VideoMode> {
+    let size = self.size().into();
+    // FIXME this is not the real refresh rate
+    // (it is guaranteed to support 32 bit color though)
+    std::iter::once(monitor::VideoMode {
+      video_mode: VideoMode {
+        size,
+        bit_depth: 32,
+        refresh_rate: 60,
+        monitor: self.clone(),
+      },
+    })
+  }
+}
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct VideoMode {
+  size: (u32, u32),
+  bit_depth: u16,
+  refresh_rate: u16,
+  monitor: MonitorHandle,
+}
+
+impl VideoMode {
+  pub fn size(&self) -> PhysicalSize<u32> {
+    self.size.into()
+  }
+
+  pub fn bit_depth(&self) -> u16 {
+    self.bit_depth
+  }
+
+  pub fn refresh_rate(&self) -> u16 {
+    self.refresh_rate
+  }
+
+  pub fn monitor(&self) -> monitor::MonitorHandle {
+    monitor::MonitorHandle {
+      inner: self.monitor.clone(),
+    }
+  }
+}
+pub fn keycode_to_scancode(_code: KeyCode) -> Option<u32> {
+  None
+}
+
+pub fn keycode_from_scancode(_scancode: u32) -> KeyCode {
+  KeyCode::Unidentified(NativeKeyCode::Unidentified)
+}
diff --git a/src/window.rs b/src/window.rs
index e43c72e..170f99e 100644
--- a/src/window.rs
+++ b/src/window.rs
@@ -1150,14 +1150,17 @@ impl Window {
   /// - **iOS / Android:** Unsupported.
   #[inline]
   pub fn set_progress_bar(&self, _progress: ProgressBarState) {
-    #[cfg(any(
-      windows,
-      target_os = "linux",
-      target_os = "dragonfly",
-      target_os = "freebsd",
-      target_os = "netbsd",
-      target_os = "openbsd",
-      target_os = "macos",
+    #[cfg(all(
+      any(
+        windows,
+        target_os = "linux",
+        target_os = "dragonfly",
+        target_os = "freebsd",
+        target_os = "netbsd",
+        target_os = "openbsd",
+        target_os = "macos",
+      ),
+      not(target_env = "ohos")
     ))]
     self.window.set_progress_bar(_progress)
   }
@@ -1197,14 +1200,17 @@ impl Window {
   /// - **iOS / Android:** Unsupported.
   #[inline]
   pub fn set_theme(&self, #[allow(unused)] theme: Option<Theme>) {
-    #[cfg(any(
-      windows,
-      target_os = "linux",
-      target_os = "dragonfly",
-      target_os = "freebsd",
-      target_os = "netbsd",
-      target_os = "openbsd",
-      target_os = "macos",
+    #[cfg(all(
+      any(
+        windows,
+        target_os = "linux",
+        target_os = "dragonfly",
+        target_os = "freebsd",
+        target_os = "netbsd",
+        target_os = "openbsd",
+        target_os = "macos",
+      ),
+      not(target_env = "ohos")
     ))]
     self.window.set_theme(theme)
   }
@@ -1225,7 +1231,10 @@ impl Window {
   ///
   /// - **iOS / Android / Windows:** Unsupported.
   pub fn set_visible_on_all_workspaces(&self, #[allow(unused)] visible: bool) {
-    #[cfg(any(target_os = "macos", target_os = "linux"))]
+    #[cfg(all(
+      any(target_os = "macos", target_os = "linux"),
+      not(target_env = "ohos")
+    ))]
     self.window.set_visible_on_all_workspaces(visible)
   }
 
